src/script-events.c

```diff
@@ -0,0 +1,343 @@
+    u64 p_t = bpf_get_current_pid_tgid();
+    u64 id = (u64)bpf_map_lookup_elem(&process_ids, &p_t);
+    if (id)
+    {
+        __builtin_memcpy(&id, (void *)id, sizeof(u64));
+        goto Pwd;
+    }
+
+    ev->id = ev_id;
+    ev->done = FALSE;
+    ev->telemetry_type = TE_SCRIPT,
+    ev->u.v.value[0] = '\0';
+    ev->u.v.truncated = FALSE;
+
+    //FILL_TELEMETRY_SYSCALL_EVENT(ev, sp);
+    //ev->u.syscall_info.ppid = ppid;
+    //ev->u.syscall_info.luid = luid;
+
+    // TODO: Do I need this?
+    //push_telemetry_event(ctx, ev);
+
+    bpf_map_update_elem(&process_ids, &p_t, &id, BPF_ANY);
+
+    // this has to be the same size as to_skip and skipped in Pwd
+    u32 count = 0;
+    char br = 0;
+    // reuse count as idx and value to save stack space
+    bpf_map_update_elem(&read_path_skip, &count, &count, BPF_ANY);
+
+Pwd:;
+
+    // the verifier complains if these are instantiated before the Pwd:; label
+    u32 to_skip = 0;
+    u32 skipped = 0;
+
+    // since index will start at zero, we can use it here
+    u64 offset = 0;
+    void *ptr = (void *)bpf_get_current_task();
+    if (read_value(ptr, CRC_TASK_STRUCT_FS, &ptr, sizeof(ptr)) < 0)
+        goto Skip;
+
+    offset = CRC_FS_STRUCT_PWD;
+    offset = (u64)bpf_map_lookup_elem(&offsets, &offset);
+    if (!offset)
+        goto Skip;
+    ptr = ptr + *(u32 *)offset; // ptr to pwd
+
+    if (read_value(ptr, CRC_PATH_DENTRY, &ptr, sizeof(ptr)) < 0)
+        goto Skip;
+
+    SET_OFFSET(CRC_DENTRY_D_NAME);
+    u32 qstr_len = *(u32 *)offset; // variable name doesn't match here, we're reusing it to preserve stack
+
+    SET_OFFSET(CRC_QSTR_NAME);
+    u32 name = qstr_len + *(u32 *)offset; // offset to name char ptr within qstr of dentry
+
+    SET_OFFSET(CRC_DENTRY_D_PARENT);
+    u32 parent = *(u32 *)offset; // offset of d_parent
+
+    u32 *_to_skip = (u32 *)bpf_map_lookup_elem(&read_path_skip, &to_skip);
+    if (_to_skip)
+    {
+        to_skip = *_to_skip;
+    }
+
+    if (to_skip != 0)
+    {
+        SKIP_PATH_N(150);
+    }
+
+Send:
+    SEND_PATH_N(10);
+
+    // update to_skip, reuse skipped as index by resetting it
+    skipped = 0;
+    bpf_map_update_elem(&read_path_skip, &skipped, &to_skip, BPF_ANY);
+    // tail call back in
+    bpf_tail_call(ctx, &tail_call_table, 0);
+
+Skip:
+    skipped = 0;
+    bpf_map_delete_elem(&read_path_skip, &skipped);
+    ev->id = id;
+    return ev;
+}
+
+SEC("kretprobe/ret_script_load")
+int kretprobe__ret_script_load(struct pt_regs *ctx)
+{
+    char br = 0;
+    u32 id = 0;
+    u32 count = 0;
+    void **bprmp = NULL;
+    void *str = NULL;
+    u64 loaded = CRC_LOADED;
+    unsigned char *bprm = NULL;
+
+    // Make sure the function succeeded
+    int ret = PT_REGS_RC(ctx);
+    if (ret < 0)
+    {
+        return 0;
+    }
+
+    // Just to be safe 0 out the structs
+    telemetry_event_t ev;
+    __builtin_memset(&ev, 0, sizeof(ev));
+
+    // Initialize some of the telemetry event
+    id = bpf_get_prandom_u32();
```

Comment:
Are you using a kprobe with matching kretprobe here? It looks like that because you check the `&load_script_map` below. In that case, the `id` should probably match across both, unless we're treating the kprobe and kretprobe as separate events.

Reply:
The kprobe doesn't actually emit any data. The kprobe is used to get the linux_bprm pointer and the kretprobe is where most of the work happens since we can check for successful completion of the function at that point. If the function fails we don't want to emit any events or clog up the cache in the plugin

Reply:
Oh neat. Good idea!